source/common/upstream/ring_hash_lb.cc

```diff
@@ -193,6 +220,54 @@ RingHashLoadBalancer::Ring::Ring(const NormalizedHostWeightVector& normalized_ho
     }
   }
 
+  if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.shard_ringhash")) {
+    /******* BEGIN: code for sharding algorithm *******/
+
+    // Find MSB bit of the first hash so we can right shift all the other hashes to create shards.
+    int msb = 0;
+    uint64_t n = ring_[0].hash_;
+    n = n / 2;
+    while (n != 0) {
+      n = n / 2;
+      msb++;
+    }
+    // Arbitrarily choosing MSB + 10 bits to shift hash to right for creating shards. The larger the
+    // shift to right, the fewer the shards. Experiment with values 9, 10, ... 30, the SHARD_SHIFT
+    // parameter.
+    rightShift += msb;
+    rightShift = (rightShift > 64) ? 64 : rightShift;
+
+    // Reserve memory for shard indices. Worst-case, every hash belongs to a different shard!
+    // The ring_shard_ container stores the start indices of the shards.
+    ring_shard_.reserve(ring_size);
+
+    // Right shift each hash and create shards of the hosts.
+    uint64_t ring_index = 0;
+    uint64_t current_shard, prev_shard = 0;
+
+    // push the first index if the ring_ isn't empty
+    if (!ring_.empty()) {
+      ring_shard_.push_back(ring_index);
+      ring_index++;
+    }
+
+    for (const auto& entry : ring_) {
+      // shifting right by 64 as 64-bit unsigned int doesn't seem to work! So, I split it.
+      current_shard = (entry.hash_ >> (rightShift - 1)) >> 1;
+
+      // If new shard found, push the index to ring_shard_ and update current_shard.
+      if (current_shard != prev_shard) {
+        prev_shard = current_shard;
+        ring_shard_.push_back(ring_index);
+      }
+      ring_index++;
+    }
+    // For the last shard, we need end and hence storing ring_size value.
+    ring_shard_.push_back(ring_size);
+
+    /******* END: code for sharding algorithm *******/
```

Comment:
Regarding documentation, see if it is useful to mention about the optimization in https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/load_balancers.html?highlight=ring%20hash#ring-hash 

I would prefer if the runtime flag is made true by default. So that others can benefit from this change. Else we wouldn't normally expect folks to explicitly enable this flag. 

Include a note in changeslog that this change optimize the execution. if this optimization is not required then this change can be reverted using the runtime flag.

Reply:
I didn't change the documentation yet. I updated the changelog: `changelogs/current.yaml`.
I changed `rightShift` to `rshift_to_shard`.